### DISCUSSION TITLE
Required: Sample request for updating the content state of an assessment

### DISCUSSION THREAD

**Discussion by Debayanpaul1593** [Sunbird-Lern/Community#161] [Query] [answered]
Hi Team, 

I need a sample request body for `/v1/content/state/update` api for updating the state of an assessment with event details. I am looking to update the progress and state of a MCQ question. Carrying forward this [discussion](https://github.com/orgs/Sunbird-inQuiry/discussions/96) .

Thanks

**Comment by BharathwajShankar** [comment]
@Debayanpaul1593 please find the curl for content state update API,

curl --location --request PATCH '{{host}}/api/course/v1/content/state/update' \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--header 'Authorization: <kong_api_key>' \
--header 'X-Authenticated-User-token: <keycloak_access_token>' \
--data '{
  "request": {
    "userId": "{{user_id}}",
    "contents": [
      {
        "contentId": "",
        "batchId": "{{batch_id}}",
        "status": 2,
        "courseId": "{{course_id}}",
        "lastAccessTime": "2020-11-23 12:58:35:179+0000"
      }
    ]
  }
}'

Please make sure you replace the exact values in the place holders.

**Reply by BharathwajShankar** [reply]
@Debayanpaul1593  for assessments you can use this curl,

curl --location --request PATCH 'https://staging.sunbirded.org/content/course/v1/content/state/update' \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--header 'Authorization:<kong_api_key>' \
--header 'X-Authenticated-User-token: <keycloak_access_token>' \
--data '{
    "request": {
        "userId": "fca2925f-1eee-4654-9177-fece3fd6afc9",
        "contents": [
            {
                "contentId": "do_2129399386333347841455",
                "batchId": "0132762583931781125",
                "status": 2,
                "courseId": "do_2132762420510146561629",
                "lastAccessTime": "2023-09-26 13:26:09:616+0530"
            }
        ],
        "assessments": [
            {
                "assessmentTs": 1695714945430,
                "batchId": "0132762583931781125",
                "courseId": "do_2132762420510146561629",
                "userId": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                "attemptId": "c8964f7315b62804d4c7a5c54ee7acc7",
                "contentId": "do_2129399386333347841455",
                "events": [
                    {
                        "eid": "ASSESS",
                        "ets": 1695714956786,
                        "ver": "3.1",
                        "mid": "ASSESS:11aa9ad19ecba2040ea04dbcf453f9ff",
                        "actor": {
                            "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                            "type": "User"
                        },
                        "context": {
                            "channel": "01269878797503692810",
                            "pdata": {
                                "id": "staging.sunbird.portal",
                                "ver": "6.0.0",
                                "pid": "sunbird-portal.contentplayer"
                            },
                            "env": "contentplayer",
                            "sid": "JCGMeGu0DsrVX7xwsqpvCiIqKqOX-D03",
                            "did": "bb6579b7819efe9c3b0164e2d368c145",
                            "cdata": [
                                {
                                    "id": "do_2132762453396848641630",
                                    "type": "course"
                                },
                                {
                                    "type": "batch",
                                    "id": "0132762583931781125"
                                },
                                {
                                    "id": "de97ce6250dad772c636dc0d66a50d62",
                                    "type": "ContentSession"
                                },
                                {
                                    "id": "41fa5b52833d4ac0ee5e8f2b6cd5fdc1",
                                    "type": "PlaySession"
                                }
                            ],
                            "rollup": {
                                "l1": "01269878797503692810",
                                "l2": "01274200593719296013305"
                            }
                        },
                        "object": {
                            "id": "do_2129399386333347841455",
                            "type": "Content",
                            "ver": "2",
                            "rollup": {
                                "l1": "do_2132762453396848641630",
                                "l2": "do_2129399386333347841455"
                            }
                        },
                        "tags": [
                            "01269878797503692810",
                            "01274200593719296013305"
                        ],
                        "edata": {
                            "item": {
                                "id": "do_2129195314499747841629",
                                "maxscore": 1,
                                "type": "mcq",
                                "exlength": 0,
                                "params": [
                                    {
                                        "1": "{\"text\":\"Ur (correct)\\n\"}"
                                    },
                                    {
                                        "2": "{\"text\":\"Sabha\\n\"}"
                                    },
                                    {
                                        "3": "{\"text\":\"Nagaram\\n\"}"
                                    },
                                    {
                                        "4": "{\"text\":\"Kurram\\n\"}"
                                    },
                                    {
                                        "answer": "{\"correct\":[\"1\"]}"
                                    }
                                ],
                                "uri": "",
                                "title": "Question 8",
                                "mmc": [],
                                "mc": [],
                                "desc": "By kiruba"
                            },
                            "index": 1,
                            "pass": "Yes",
                            "score": 1,
                            "resvalues": [
                                {
                                    "1": "{\"text\":\"Ur (correct)\\n\"}"
                                }
                            ],
                            "duration": 2
                        }
                    },
                    {
                        "eid": "ASSESS",
                        "ets": 1695714958198,
                        "ver": "3.1",
                        "mid": "ASSESS:8becc596aed0bba78e5d52afe6bac97f",
                        "actor": {
                            "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                            "type": "User"
                        },
                        "context": {
                            "channel": "01269878797503692810",
                            "pdata": {
                                "id": "staging.sunbird.portal",
                                "ver": "6.0.0",
                                "pid": "sunbird-portal.contentplayer"
                            },
                            "env": "contentplayer",
                            "sid": "JCGMeGu0DsrVX7xwsqpvCiIqKqOX-D03",
                            "did": "bb6579b7819efe9c3b0164e2d368c145",
                            "cdata": [
                                {
                                    "id": "do_2132762453396848641630",
                                    "type": "course"
                                },
                                {
                                    "type": "batch",
                                    "id": "0132762583931781125"
                                },
                                {
                                    "id": "de97ce6250dad772c636dc0d66a50d62",
                                    "type": "ContentSession"
                                },
                                {
                                    "id": "41fa5b52833d4ac0ee5e8f2b6cd5fdc1",
                                    "type": "PlaySession"
                                }
                            ],
                            "rollup": {
                                "l1": "01269878797503692810",
                                "l2": "01274200593719296013305"
                            }
                        },
                        "object": {
                            "id": "do_2129399386333347841455",
                            "type": "Content",
                            "ver": "2",
                            "rollup": {
                                "l1": "do_2132762453396848641630",
                                "l2": "do_2129399386333347841455"
                            }
                        },
                        "tags": [
                            "01269878797503692810",
                            "01274200593719296013305"
                        ],
                        "edata": {
                            "item": {
                                "id": "do_2129195303901757441628",
                                "maxscore": 1,
                                "type": "mcq",
                                "exlength": 0,
                                "params": [
                                    {
                                        "1": "{\"text\":\"Narasimhavarman II (correct)\\n\"}"
                                    },
                                    {
                                        "2": "{\"text\":\"Ravivarman\\n\"}"
                                    },
                                    {
                                        "3": "{\"text\":\"Mahendravarman\\n\"}"
                                    },
                                    {
                                        "4": "{\"text\":\"Devavarman\\n\"}"
                                    },
                                    {
                                        "answer": "{\"correct\":[\"1\"]}"
                                    }
                                ],
                                "uri": "",
                                "title": "Question 7",
                                "mmc": [],
                                "mc": [],
                                "desc": "By kiruba"
                            },
                            "index": 2,
                            "pass": "Yes",
                            "score": 1,
                            "resvalues": [
                                {
                                    "1": "{\"text\":\"Narasimhavarman II (correct)\\n\"}"
                                }
                            ],
                            "duration": 1
                        }
                    },
                    {
                        "eid": "ASSESS",
                        "ets": 1695714959152,
                        "ver": "3.1",
                        "mid": "ASSESS:adcf0da641ae451f1e9ec36e8494b8fe",
                        "actor": {
                            "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                            "type": "User"
                        },
                        "context": {
                            "channel": "01269878797503692810",
                            "pdata": {
                                "id": "staging.sunbird.portal",
                                "ver": "6.0.0",
                                "pid": "sunbird-portal.contentplayer"
                            },
                            "env": "contentplayer",
                            "sid": "JCGMeGu0DsrVX7xwsqpvCiIqKqOX-D03",
                            "did": "bb6579b7819efe9c3b0164e2d368c145",
                            "cdata": [
                                {
                                    "id": "do_2132762453396848641630",
                                    "type": "course"
                                },
                                {
                                    "type": "batch",
                                    "id": "0132762583931781125"
                                },
                                {
                                    "id": "de97ce6250dad772c636dc0d66a50d62",
                                    "type": "ContentSession"
                                },
                                {
                                    "id": "41fa5b52833d4ac0ee5e8f2b6cd5fdc1",
                                    "type": "PlaySession"
                                }
                            ],
                            "rollup": {
                                "l1": "01269878797503692810",
                                "l2": "01274200593719296013305"
                            }
                        },
                        "object": {
                            "id": "do_2129399386333347841455",
                            "type": "Content",
                            "ver": "2",
                            "rollup": {
                                "l1": "do_2132762453396848641630",
                                "l2": "do_2129399386333347841455"
                            }
                        },
                        "tags": [
                            "01269878797503692810",
                            "01274200593719296013305"
                        ],
                        "edata": {
                            "item": {
                                "id": "do_2129174037204910081236",
                                "maxscore": 1,
                                "type": "mcq",
                                "exlength": 0,
                                "params": [
                                    {
                                        "1": "{\"text\":\"KMnO4 is an oxidising agent, it oxidises FeSO4 (correct)\\n\"}"
                                    },
                                    {
                                        "2": "{\"text\":\"FeSO4 acts as an oxidising agent and oxidises KMnO4\\n\"}"
                                    },
                                    {
                                        "3": "{\"text\":\"The colour disappears due to dilution, no reaction is involved\\n\"}"
                                    },
                                    {
                                        "4": "{\"text\":\"KMnO4 is an unstable compound and decomposes in the presence of FeSO4 to a colourless compound\\n\"}"
                                    },
                                    {
                                        "answer": "{\"correct\":[\"1\"]}"
                                    }
                                ],
                                "uri": "",
                                "title": "Explore Question 6",
                                "mmc": [],
                                "mc": [],
                                "desc": "By kiruba"
                            },
                            "index": 3,
                            "pass": "No",
                            "score": 0,
                            "resvalues": [
                                {
                                    "2": "{\"text\":\"FeSO4 acts as an oxidising agent and oxidises KMnO4\\n\"}"
                                }
                            ],
                            "duration": 1
                        }
                    },
                    {
                        "eid": "ASSESS",
                        "ets": 1695714960027,
                        "ver": "3.1",
                        "mid": "ASSESS:e9e4d066395f3d1203740548f5910d2b",
                        "actor": {
                            "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                            "type": "User"
                        },
                        "context": {
                            "channel": "01269878797503692810",
                            "pdata": {
                                "id": "staging.sunbird.portal",
                                "ver": "6.0.0",
                                "pid": "sunbird-portal.contentplayer"
                            },
                            "env": "contentplayer",
                            "sid": "JCGMeGu0DsrVX7xwsqpvCiIqKqOX-D03",
                            "did": "bb6579b7819efe9c3b0164e2d368c145",
                            "cdata": [
                                {
                                    "id": "do_2132762453396848641630",
                                    "type": "course"
                                },
                                {
                                    "type": "batch",
                                    "id": "0132762583931781125"
                                },
                                {
                                    "id": "de97ce6250dad772c636dc0d66a50d62",
                                    "type": "ContentSession"
                                },
                                {
                                    "id": "41fa5b52833d4ac0ee5e8f2b6cd5fdc1",
                                    "type": "PlaySession"
                                }
                            ],
                            "rollup": {
                                "l1": "01269878797503692810",
                                "l2": "01274200593719296013305"
                            }
                        },
                        "object": {
                            "id": "do_2129399386333347841455",
                            "type": "Content",
                            "ver": "2",
                            "rollup": {
                                "l1": "do_2132762453396848641630",
                                "l2": "do_2129399386333347841455"
                            }
                        },
                        "tags": [
                            "01269878797503692810",
                            "01274200593719296013305"
                        ],
                        "edata": {
                            "item": {
                                "id": "do_2129174030324285441235",
                                "maxscore": 1,
                                "type": "mcq",
                                "exlength": 0,
                                "params": [
                                    {
                                        "1": "{\"text\":\"crystals immediately melt\\n\"}"
                                    },
                                    {
                                        "2": "{\"text\":\"a brown residue is left (correct) \\n\"}"
                                    },
                                    {
                                        "3": "{\"text\":\"white fumes appear in the test tube\\n\"}"
                                    },
                                    {
                                        "4": "{\"text\":\"a yellow residue is left\\n\"}"
                                    },
                                    {
                                        "answer": "{\"correct\":[\"2\"]}"
                                    }
                                ],
                                "uri": "",
                                "title": "Explore Question 5",
                                "mmc": [],
                                "mc": [],
                                "desc": "By kiruba"
                            },
                            "index": 4,
                            "pass": "Yes",
                            "score": 1,
                            "resvalues": [
                                {
                                    "2": "{\"text\":\"a brown residue is left (correct) \\n\"}"
                                }
                            ],
                            "duration": 1
                        }
                    },
                    {
                        "eid": "ASSESS",
                        "ets": 1695714962218,
                        "ver": "3.1",
                        "mid": "ASSESS:c9045a0844652e78a112a7ca7f9159e2",
                        "actor": {
                            "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                            "type": "User"
                        },
                        "context": {
                            "channel": "01269878797503692810",
                            "pdata": {
                                "id": "staging.sunbird.portal",
                                "ver": "6.0.0",
                                "pid": "sunbird-portal.contentplayer"
                            },
                            "env": "contentplayer",
                            "sid": "JCGMeGu0DsrVX7xwsqpvCiIqKqOX-D03",
                            "did": "bb6579b7819efe9c3b0164e2d368c145",
                            "cdata": [
                                {
                                    "id": "do_2132762453396848641630",
                                    "type": "course"
                                },
                                {
                                    "type": "batch",
                                    "id": "0132762583931781125"
                                },
                                {
                                    "id": "de97ce6250dad772c636dc0d66a50d62",
                                    "type": "ContentSession"
                                },
                                {
                                    "id": "41fa5b52833d4ac0ee5e8f2b6cd5fdc1",
                                    "type": "PlaySession"
                                }
                            ],
                            "rollup": {
                                "l1": "01269878797503692810",
                                "l2": "01274200593719296013305"
                            }
                        },
                        "object": {
                            "id": "do_2129399386333347841455",
                            "type": "Content",
                            "ver": "2",
                            "rollup": {
                                "l1": "do_2132762453396848641630",
                                "l2": "do_2129399386333347841455"
                            }
                        },
                        "tags": [
                            "01269878797503692810",
                            "01274200593719296013305"
                        ],
                        "edata": {
                            "item": {
                                "id": "do_2129174021075271681233",
                                "maxscore": 1,
                                "type": "mcq",
                                "exlength": 0,
                                "params": [
                                    {
                                        "1": "{\"text\":\"Carbohydrates\\n\"}"
                                    },
                                    {
                                        "2": "{\"text\":\"Small globules\\n\"}"
                                    },
                                    {
                                        "3": "{\"text\":\"Amino acids (correct)\\n\"}"
                                    },
                                    {
                                        "4": "{\"text\":\"Starch\\n\"}"
                                    },
                                    {
                                        "answer": "{\"correct\":[\"3\"]}"
                                    }
                                ],
                                "uri": "",
                                "title": "Explore Question 4",
                                "mmc": [],
                                "mc": [],
                                "desc": "By kiruba"
                            },
                            "index": 5,
                            "pass": "No",
                            "score": 0,
                            "resvalues": [
                                {
                                    "2": "{\"text\":\"Small globules\\n\"}"
                                }
                            ],
                            "duration": 2
                        }
                    },
                    {
                        "eid": "ASSESS",
                        "ets": 1695714966932,
                        "ver": "3.1",
                        "mid": "ASSESS:75987a0b07e3684f184dafc413570cb9",
                        "actor": {
                            "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
                            "type": "User"
                        },
                        "context": {
                            "channel": "01269878797503692810",
                            "pdata": {
                                "id": "staging.sunbird.portal",
                                "ver": "6.0.0",
                                "pid": "sunbird-portal.contentplayer"
                            },
                            "env": "contentplayer",
                            "sid": "JCGMeGu0DsrVX7xwsqpvCiIqKqOX-D03",
                            "did": "bb6579b7819efe9c3b0164e2d368c145",
                            "cdata": [
                                {
                                    "id": "do_2132762453396848641630",
                                    "type": "course"
                                },
                                {
                                    "type": "batch",
                                    "id": "0132762583931781125"
                                },
                                {
                                    "id": "de97ce6250dad772c636dc0d66a50d62",
                                    "type": "ContentSession"
                                },
                                {
                                    "id": "41fa5b52833d4ac0ee5e8f2b6cd5fdc1",
                                    "type": "PlaySession"
                                }
                            ],
                            "rollup": {
                                "l1": "01269878797503692810",
                                "l2": "01274200593719296013305"
                            }
                        },
                        "object": {
                            "id": "do_2129399386333347841455",
                            "type": "Content",
                            "ver": "2",
                            "rollup": {
                                "l1": "do_2132762453396848641630",
                                "l2": "do_2129399386333347841455"
                            }
                        },
                        "tags": [
                            "01269878797503692810",
                            "01274200593719296013305"
                        ],
                        "edata": {
                            "item": {
                                "id": "do_2129174014605885441232",
                                "maxscore": 1,
                                "type": "mcq",
                                "exlength": 0,
                                "params": [
                                    {
                                        "1": "{\"text\":\"Absorbing light (correct)\\n\"}"
                                    },
                                    {
                                        "2": "{\"text\":\"Breaking down water molecule\\n\"}"
                                    },
                                    {
                                        "3": "{\"text\":\"No function\\n\"}"
                                    },
                                    {
                                        "4": "{\"text\":\"Reduction of CO2\\n\"}"
                                    },
                                    {
                                        "answer": "{\"correct\":[\"1\"]}"
                                    }
                                ],
                                "uri": "",
                                "title": "Explore Question 3",
                                "mmc": [],
                                "mc": [],
                                "desc": "By kiruba"
                            },
                            "index": 6,
                            "pass": "No",
                            "score": 0,
                            "resvalues": [
                                {
                                    "2": "{\"text\":\"Breaking down water molecule\\n\"}"
                                }
                            ],
                            "duration": 5
                        }
                    }
                ]
            }
        ]
    }
}'

**Comment by Debayanpaul1593** [comment]
Hi @BharathwajShankar thank you for the quick response. Currently I am working on recording the answers to assessments and updating the completion status for the same. However I intend to work with minimum data as we are still in beginning stages and do not have complete understanding of sunbird specifications. I have 3 questions following this:
1. As per [telemetry specification](https://telemetry.sunbird.org/learn/specification#telemetry-v3-event-structure) the minimum event structure after removing the optional fields can be:
```JSON
{
  "eid": "ASSESS",
  "ets": "1695714962218",
  "ver": "3.1",
  "mid": "ASSESS:c9045a0844652e78a112a7ca7f9159e2",
  "actor": {
    "id": "fca2925f-1eee-4654-9177-fece3fd6afc9",
    "type": "User"
  },
  "context": {
    "channel": "01269878797503692810",
    "env": "contentplayer"
  },
  "edata": {
    "item": {
      "id": "do_2129174014605885441232",
      "maxscore": 1,
      "type": "mcq",
      "exlength": 0,
      "params": [
        {
          "1": "{\"text\":\"Absorbing light (correct)\n\"}"
        },
        {
          "2": "{\"text\":\"Breaking down water molecule\n\"}"
        },
        {
          "3": "{\"text\":\"No function\n\"}"
        },
        {
          "4": "{\"text\":\"Reduction of CO2\n\"}"
        },
        {
          "answer": "{\"correct\":[\"1\"]}"
        }
      ],
      "uri": "",
      "title": "Explore Question 3",
      "mmc": [],
      "mc": [],
      "desc": "By kiruba"
    },
    "index": 6,
    "pass": "No",
    "score": 0,
    "resvalues": [
      {
        "2": "{\"text\":\"Breaking down water molecule\n\"}"
      }
    ],
    "duration": 5
  }
}
```
Is this sufficient to record the answer to an event and update the status/progress of the assessment?
2. It seems the important data points are being sent in the **edata** object. But there is no explanation for the same in the telemetry specification. Could you please provide understanding of the same?
3. In the sunbird specification there is another event with a similar description as **ASSESS** event type called **RESPONSE**. Both seem to be related to capturing/recording question/assessment answers. Can you explain the difference and the reason for using **ASSESS** type and not **RESPONSE** type?

**Reply by kumarks1122** [reply]
@Debayanpaul1593 : Yes .. The mentioned sample event has sufficient data for processing the event. ASSESS event is primarily meant for getting the user response details from the assessment type of content. Also the flink job(assessment-aggregator) updating the Cassandra table with user response requires ASSESS type of event.
Kindly find the telemetry spec for ASSESS event from the below links
https://telemetry.sunbird.org/learn/specification#assess
https://telemetry.sunbird.org/learn/v3_event_details#assess

**Comment by Debayanpaul1593** [comment]
Hi @BharathwajShankar, 
In the request is the content item inside assessments array i.e., with id do_2129399386333347841455 a Questionset? And are the event datas inside the events array all the questions?

**Reply by kumarks1122** [reply, marked as answer]
@Debayanpaul1593 do_2129399386333347841455 is an assessment. And assessments array contains the response of each question in that assessment.
`FYI:` Even QuestionSet also can be implemented to send ASSESS events from portal side.

**Reply by mohitga** [reply]
@Debayanpaul1593  marking the thread as answered, Please feel free to open the thread again if you any query on the same.
